### PR TITLE
[flutter_tools] remove all pub caching logic

### LINKED ISSUE #66777: pub reruns can lead to unecessary rebuilds of application due to "generated": <Date> field in package_config
Proposed solution: generate a new file from the package config that is a subset of the contents of this file and perform all diffing on this one instead.

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

There have been some more additional reports of a missing 'package:characters' import after upgrading flutter. This has me concerned that our pub caching logic is incorrect. Instead of the tool attempting to guess when pub should be run, always delegate to pub.

Also takes an opportunity to fix the kernel snapshot depending on the .packages or package_config. Due to the generated: date field this causes extra rebuilds. Instead when pub get is run, write out an additional file with just the package contents and version.

Fixes https://github.com/flutter/flutter/issues/66777

### CODE REVIEW

File: packages/flutter_tools/lib/src/build_system/targets/android.dart [hunk 1]
```diff
@@ -192,7 +192,6 @@ class AndroidAot extends AotElfBase {
   List<Source> get inputs => <Source>[
     const Source.pattern('{FLUTTER_ROOT}/packages/flutter_tools/lib/src/build_system/targets/android.dart'),
     const Source.pattern('{BUILD_DIR}/app.dill'),
-    const Source.pattern('{PROJECT_DIR}/.packages'),
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/build_system/targets/android.dart:195]: Doesn't need to depend on this

File: packages/flutter_tools/lib/src/build_system/targets/common.dart [hunk 2]
```diff
@@ -171,7 +175,7 @@ class KernelSnapshot extends Target {
 
   @override
   List<Source> get inputs => const <Source>[
-    Source.pattern('{PROJECT_DIR}/.packages'),
+    Source.pattern('{PROJECT_DIR}/.dart_tool/package_config_subset'),
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/build_system/targets/common.dart:178]: generated by pub now

File: packages/flutter_tools/lib/src/dart/pub.dart [hunk 3]
```diff
@@ -168,92 +162,57 @@ class _DefaultPub implements Pub {
...
-
-    if (!checkLastModified || _shouldRunPubGet(
-      pubSpecYaml: pubSpecYaml,
-      packageConfigFile: packageConfigFile,
-    )) {
-      final String command = upgrade ? 'upgrade' : 'get';
-      final Status status = _logger.startProgress(
-        'Running "flutter pub $command" in ${_fileSystem.path.basename(directory)}...',
-        timeout: const TimeoutConfiguration().slowOperation,
+    final String command = upgrade ? 'upgrade' : 'get';
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/dart/pub.dart:174]: Only change is the removal of the conditional run logic

File: packages/flutter_tools/lib/src/dart/pub.dart [hunk 4]
```diff
@@ -458,13 +400,31 @@ class _DefaultPub implements Pub {
...
+  /// if [generateSyntheticPackage] is true then insert flutter_gen synthetic
+  /// package into the package configuration.
+  Future<void> _updatePackageConfig(
+    File packageConfigFile,
+    Directory generatedDirectory,
+    bool generateSyntheticPackage,
+  ) async {
+    final PackageConfig packageConfig = await loadPackageConfigWithLogging(packageConfigFile, logger: _logger);
+
+    packageConfigFile.parent
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/dart/pub.dart:417]: This is where the subset of package contents are written out, avoids the generated by and generated date fields

File: packages/flutter_tools/lib/src/runner/flutter_command.dart [hunk 5]
```diff
@@ -934,9 +934,9 @@ abstract class FlutterCommand extends Command<void> {
       // First always update universal artifacts, as some of these (e.g.
       // ios-deploy on macOS) are required to determine `requiredArtifacts`.
       await globals.cache.updateAll(<DevelopmentArtifact>{DevelopmentArtifact.universal});
-
       await globals.cache.updateAll(await requiredArtifacts);
     }
+    Cache.releaseLock();
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/runner/flutter_command.dart:939]: Lock can be released earlier

**Comment by jmagman** [comment]
I feel like I've seen package issues when I downgrade flutter from master to stable after I delete my pub cache.  I think it was that the master version of flutter used something like `character 1.1.0-nullsafety.3` package (first one it tries it `character` so that was always the error) which requires dart `>=2.10.0-110`.  And then I downgrade `flutter` to `stable` and the project pubspec is fine with the character package range so `pub get` doesn't try to download an older version of the package, but it complains that the dart version is too old.  I think a `flutter clean` got me out of it.

If I see it again I'll look into it, it's a vague recollection.

File: packages/flutter_tools/lib/src/build_system/targets/common.dart [hunk 6]
```diff
@@ -161,8 +161,12 @@ class ReleaseCopyFlutterBundle extends CopyFlutterBundle {
   List<Target> get dependencies => const <Target>[];
 }
 
-
 /// Generate a snapshot of the dart code used in the program.
+///
+/// Note that this target depends on the `.dart_tool/package_config.json` file
```
Comment (jmagman) [inline packages/flutter_tools/lib/src/build_system/targets/common.dart:166]: Why isn't it listed as an input?

File: packages/flutter_tools/test/general.shard/build_system/targets/common_test.dart [hunk 7]
```diff
@@ -91,7 +93,7 @@ void main() {
         '--aot',
         '--tfa',
         '--packages',
-        '/.packages',
+       '/.dart_tool/package_config.json',
```
Comment (jmagman) [inline packages/flutter_tools/test/general.shard/build_system/targets/common_test.dart:96]: Missing a leading space, or GitHub is tricking me.

File: packages/flutter_tools/test/general.shard/build_system/targets/common_test.dart [hunk 8]
```diff
@@ -125,7 +129,7 @@ void main() {
...
```
Comment (jmagman) [inline packages/flutter_tools/test/general.shard/build_system/targets/common_test.dart:132]: Space

**Review by jmagman — COMMENTED** [review]
> Also takes an opportunity to fix the kernel snapshot depending on the .packages or package_config. 

Maybe you should split this out into its own review.  If we decide to revert the no-caching change it would be good to still have this improvement.

Really a shame we are removing the caching when we don't understand why it's failing--it can take kind of awhile to run even when nothing is updated.

**Comment by jonahwilliams** [comment]
> Maybe you should split this out into its own review. If we decide to revert the no-caching change it would be good to still have this improvement.

Without the caching change our tests that verify repeated runs of build apk/ios are cached correctly fail

**Comment by jonahwilliams** [comment]
> Really a shame we are removing the caching when we don't understand why it's failing--it can take kind of awhile to run even when nothing is updated.

I don't quite agree - we know that it is failing because our checks are not exhaustive, so we have two choices - duplicate the caching code from pub exactly, tests and all - or skip the checks and let pub decide.

**Comment by jonahwilliams** [comment]
Oops, I guess I could do the pub caching change first and then remove the caching

**Comment by jonahwilliams** [comment]
Rebased after package subset changes landed, PTAL @jmagman

**Review by jmagman — APPROVED** [review]
The code change LGTM (and thanks for splitting it out).  Do we have any metrics to know how this will regress timing?

**Comment by jonahwilliams** [comment]
Yes, there is a flutter test perf benchmark which measure end to end completion of a single test case (the hello world example). For other commands like run it will not matter as much since those commands already take many seconds the connect.

**Comment by liyuqian** [comment]
@jmagman : here's s specific metric that would capture the time change of this PR https://flutter-flutter-perf.skia.org/e/?begin=1601510523&end=1602169767&keys=X6212caff01d08c039d5087d22c8754ea&requestType=0&xbaroffset=21843

**Pull request merged**